src/rebar_compiler_dag.erl

```diff
@@ -484,3 +484,63 @@ is_dirty(G) ->
 %% vertices, clear the flag before serializing it.
 clear_dirty(G) ->
     digraph:del_vertex(G, '$r3_dirty_bit').
+
+%% Resolve all the dependencies of a .hrl file transitively. Use a cache to
+%% memoize the resolution.
+resolve_hrl_dependencies(Name, Cache, G) ->
+    case is_hrl_file(Name) of
+        false ->
+            {[Name], Cache};
+        true ->
+            case lookup_hrl(Name, Cache) of
+                {ok, Deps} -> {Deps, Cache};
+                error ->
+                    {Deps, NewCache} = resolve_full_hrl(Name, Cache, G),
+                    {Deps, add_hrl(Name, Deps, NewCache)}
+            end
+    end.
+
+is_hrl_file(Name) ->
+    filename:extension(Name) == ".hrl".
```

Comment:
we should rely on the similar app-configurable logic here. The rebar3 compiler is pluggable and can be used by non-erlang languages (I have my own blog compile markdown files to HTML using this code as a plugin).

As such, you'll find the code in this file does not rely on ".hrl" or other erlang-specific extensions. The 3 categories are:

1. `SrcExt` (i.e. `.erl`)
2. `ArtifactExt` (i.e. `.beam`)
3. Other file (i.e. `.hrl`)

The thing here is that 'Other' files are assumed to be headers. The annoying bit is that we can't be fully sure that headers are _not_ going to be significant in other languages, but if we keep the current semantics and just memoize it, it should be fine as long as we adjust the matching criteria to be done by the list above.

I'm thinking that it's fine to do this short-circuit for Erlang specifically _if we change semantics_, but we should frame and gate it properly, keeping in mind this module is generally language-agnostic.

Reply:
I have changed it to check that a file is not Src and not Artifact to be expanded this way, was that what you had in mind or adding a third category which would be "resolvable extension"?